File: 프로그래머스/lv2/이재표/개인정보 수집 유효기간.java
import java.util.*;
class Solution {
    public static int[] solution(String today, String[] terms, String[] privacies) {
        List<Integer> list = new ArrayList<>();
        Map<String, String> termMap = new HashMap<>();
        for (int i = 0; i < terms.length; i++) {
            String[] term = terms[i].split(" ");
            termMap.put(term[0], term[1]);
        }
        for (int i = 0; i < privacies.length; i++) {
            String[] privacy = privacies[i].split(" ");
            String endDate = calEndDate(privacy[0], termMap.get(privacy[1]));
            if (canNotHavePrivacy(today, endDate) >= 0) {
                list.add(i + 1);
            }
        }
        int[] answer = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        return answer;
    }

    public static int canNotHavePrivacy(String stand, String date) {
        return stand.compareTo(date);
    }

    public static String calEndDate(String date, String month) {
        StringBuilder sb = new StringBuilder();
        String[] splitDate = date.split("\\.");
        Integer dm = Integer.parseInt(splitDate[1]);
        Integer m = Integer.parseInt(month);
        dm = dm + m;
        int plusY = 0;
        while (dm > 12) {
            dm -= 12;
            plusY++;
        }
        Integer dy = Integer.parseInt(splitDate[0]) + plusY;
        sb.append(String.valueOf(dy));
        sb.append(".");
        sb.append(String.format("%02d", dm));
        sb.append(".");
        sb.append(splitDate[2]);
        return sb.toString();
    }

}

File: 프로그래머스/lv2/이하은/개인정보 수집 유효기간.kt
class Solution {
    fun calcTotalDays(temp: List<String>): Int {
        return temp[0].toInt() * 12 * 28 + temp[1].toInt() * 28 + temp[2].toInt()
    }
    
    fun solution(today: String, terms: Array<String>, privacies: Array<String>): List<Int> {
        val answer = mutableListOf<Int>()
        val todayDays = calcTotalDays(today.split("."))
        
        val hash = hashMapOf<String, Int>() 
        for(term in terms){
            val temp = term.split(" ")
            val type = temp[0]
            val month = temp[1].toInt()
            hash[type] = month
        }
        
        privacies.forEachIndexed { idx, privacy -> 
            val temp = privacy.split(" ")
            val collectDays = calcTotalDays(temp[0].split("."))
            val termType = temp[1]
            
            // 수집 일자 + 약관의 유효 기간
            val month = hash[termType]!!
            val expireDays = collectDays + (month * 28) - 1

            // 파기할 개인정보 번호 저장
            if(todayDays > expireDays){
                answer.add(idx + 1)
            }
        }
        
        return answer
    }
}
